// neonctl/psql/describe/versionGate.js
/**
 * Server-version comparison helpers.
 *
 * PostgreSQL encodes server_version_num as MMmmpp (e.g. 9.6.0 -> 90600,
 * 10.0 -> 100000, 17.0 -> 170000). See `PG_VERSION_NUM` in upstream.
 *
 * These helpers replace the inline `pset.sversion >= N` checks that pepper
 * `describe.c`. Keeping them in one place lets a future WP centrally adjust
 * minimum-supported-version policy.
 */
/**
 * Encode major[.minor] into the server_version_num integer form.
 * For PG >= 10 the minor component is ignored (single-component versioning).
 *
 * `major` may also be an already-encoded `ServerVersion` (>= 10000), in which
 * case it is returned unchanged — this lets callers pass either the bare
 * major number (e.g. `11`) or one of the `PG_*` constants (e.g. `PG_11`).
 */
const encode = (major, minor = 0) => {
    if (major >= 10000)
        return major; // already encoded
    if (major >= 10)
        return major * 10000;
    return major * 10000 + minor * 100;
};
export const serverAtLeast = (actual, major, minor = 0) => actual >= encode(major, minor);
export const serverLess = (actual, major, minor = 0) => actual < encode(major, minor);
export const PG_9_0 = 90000;
export const PG_9_1 = 90100;
export const PG_9_2 = 90200;
export const PG_9_3 = 90300;
export const PG_9_4 = 90400;
export const PG_9_5 = 90500;
export const PG_9_6 = 90600;
export const PG_10 = 100000;
export const PG_11 = 110000;
export const PG_12 = 120000;
export const PG_13 = 130000;
export const PG_14 = 140000;
export const PG_15 = 150000;
export const PG_16 = 160000;
export const PG_17 = 170000;
export const PG_18 = 180000;
