// neonctl/psql/wire/notify.js
/**
 * Async multiplexer for NOTICE / NOTIFY (WP-02).
 *
 * `PgConnection` keeps a single NoticeMultiplexer that all incoming
 * NoticeResponse and NotificationResponse messages flow through. Subscribers
 * (e.g. the REPL, query runner, an external observer) register handlers and
 * receive a disposer function — the same shape upstream pg uses.
 *
 * We don't use Node's `EventEmitter` because:
 *   - The Connection interface (frozen WP-00) returns disposers from
 *     `onNotice` / `onNotification`, not subscription objects.
 *   - We want exceptions in one handler to be isolated from the rest.
 */
export class NoticeMultiplexer {
    constructor() {
        this.noticeHandlers = new Set();
        this.notificationHandlers = new Set();
        /** Last error thrown by a handler. Exposed for diagnostics / tests. */
        this.lastHandlerError = undefined;
    }
    /** Subscribe to NoticeResponse. Returns a disposer. */
    onNotice(handler) {
        this.noticeHandlers.add(handler);
        return () => this.noticeHandlers.delete(handler);
    }
    /** Subscribe to NotificationResponse (LISTEN/NOTIFY). Returns a disposer. */
    onNotification(handler) {
        this.notificationHandlers.add(handler);
        return () => this.notificationHandlers.delete(handler);
    }
    emit(notice) {
        for (const h of this.noticeHandlers) {
            // Isolate handler failures: one bad subscriber must not break the
            // whole connection. We swallow + record via `lastHandlerError` so a
            // test can observe it without us pulling in a logger.
            try {
                h(notice);
            }
            catch (err) {
                this.lastHandlerError = err;
            }
        }
    }
    emitNotification(channel, payload, pid) {
        for (const h of this.notificationHandlers) {
            try {
                h(channel, payload, pid);
            }
            catch (err) {
                this.lastHandlerError = err;
            }
        }
    }
    /** Drop every subscriber. Called by Connection.close(). */
    clear() {
        this.noticeHandlers.clear();
        this.notificationHandlers.clear();
    }
}
