// neonctl/functions_api.js
import { ContentType } from '@neondatabase/api-client';
const functionsPath = (projectId, branchId) => `/projects/${encodeURIComponent(projectId)}/branches/${encodeURIComponent(branchId)}/functions`;
export const listFunctions = async (apiClient, projectId, branchId) => {
    const { data } = await apiClient.request({
        path: functionsPath(projectId, branchId),
        method: 'GET',
        secure: true,
        format: 'json',
    });
    return data.functions;
};
export const getFunction = async (apiClient, projectId, branchId, slug) => {
    const { data } = await apiClient.request({
        path: `${functionsPath(projectId, branchId)}/${encodeURIComponent(slug)}`,
        method: 'GET',
        secure: true,
        format: 'json',
    });
    return data.function;
};
export const deleteFunction = async (apiClient, projectId, branchId, slug) => {
    await apiClient.request({
        path: `${functionsPath(projectId, branchId)}/${encodeURIComponent(slug)}`,
        method: 'DELETE',
        secure: true,
    });
};
export const createDeployment = async (apiClient, projectId, branchId, slug, params) => {
    const form = new FormData();
    form.append('zip', new Blob([params.zip]), 'bundle.zip');
    form.append('memory_mib', String(params.memoryMib));
    form.append('runtime', params.runtime);
    if (params.environment)
        form.append('environment', params.environment);
    // The deploy POST returns an operation the CLI cannot poll; the body is
    // ignored. We only need the request to succeed.
    await apiClient.request({
        path: `${functionsPath(projectId, branchId)}/${encodeURIComponent(slug)}/deployments`,
        method: 'POST',
        type: ContentType.FormData,
        body: form,
        secure: true,
    });
};
